### PR TITLE
Robot suit hard-del / exited cleanup

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

Constructed cyborgs seemed to hard delete pretty consistently due to their robot suit having poor reference handing

Goes through and better implements exited for cyborgs and robot suits

Also a hard delete with forced AIs that may have resulted in some metagaming 

Also fixes 1 nodrop check for hats

## Why It's Good For The Game

Hard deletes are bad

## Changelog

:cl: Melbert
fix: Fixed some hard deletes involving constructed cyborgs
fix: Cyborgs with nodrop hats no longer lose them on tip
/:cl:

### CODE REVIEW

File: code/game/objects/items/robot/robot_parts.dm [hunk 1]
```diff
@@ -21,8 +21,13 @@
 	var/obj/item/bodypart/head/robot/head = null
 	/// Forced name of the cyborg
 	var/created_name = ""
+
 	/// Forced master AI of the cyborg
 	var/mob/living/silicon/ai/forced_ai
+	/// The name of the AI being forced, tracked separetly to above
```
Comment (san7890) [inline code/game/objects/items/robot/robot_parts.dm:27]:
```suggestion
	/// The name of the AI being forced, tracked separately to above
```

File: code/game/objects/items/robot/robot_parts.dm [hunk 2]
```diff
@@ -417,16 +439,47 @@
 			log_silicon("[key_name(user)] has [aisync ? "enabled" : "disabled"] the AI sync for a cyborg shell at [loc_name(user)]")
 			return TRUE
 		if("set_ai")
-			var/selected_ai = select_active_ai(user, z)
-			if(!in_range(src, user) && loc != user)
-				return
-			if(!selected_ai)
+			if(length(active_ais(check_mind = FALSE, z = z)) <= 0)
```
Comment (san7890) [inline code/game/objects/items/robot/robot_parts.dm:442]: are we getting a negative `length()` somehow? why not just do a `if(!length(thing))`? or just `if(length(thing) == 0)`? the `<` is confusing

**Review by san7890 — CHANGES_REQUESTED** [review]
nits

File: code/game/objects/items/robot/robot_parts.dm (hunk 2 above)
Comment (MrMelbert) [reply code/game/objects/items/robot/robot_parts.dm:442]: i just do it because i find it more readable that `!length`

Comment (san7890) [reply code/game/objects/items/robot/robot_parts.dm:442]: that's fine then but i just have no clue why the code would insinuate that length could ever be less than zero

**Pull request merged**